compiler/rustc_trait_selection/src/error_reporting/traits/suggestions.rs

```diff
@@ -5127,6 +5127,73 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             _ => {}
         }
     }
+    pub(crate) fn suggest_impl_similarly_named_trait(
+        &self,
+        err: &mut Diag<'_>,
+        obligation: &PredicateObligation<'tcx>,
+        trait_predicate: ty::PolyTraitPredicate<'tcx>,
+    ) {
+        let trait_def_id = trait_predicate.def_id();
+        let trait_name = self.tcx.item_name(trait_def_id);
+
+        let trait_has_same_params = |other_trait_def_id: DefId| -> bool {
+            let trait_generics = self.tcx.generics_of(trait_def_id);
+            let other_trait_generics = self.tcx.generics_of(other_trait_def_id);
+
+            if trait_generics.count() != other_trait_generics.count() {
+                return false;
+            }
+            trait_generics.own_params.iter().zip(other_trait_generics.own_params.iter()).all(
+                |(a, b)| {
+                    a.name == b.name
+                        && a.index == b.index
+                        && ((matches!(a.kind, ty::GenericParamDefKind::Type { .. })
+                            && matches!(b.kind, ty::GenericParamDefKind::Type { .. }))
+                            || (matches!(a.kind, ty::GenericParamDefKind::Lifetime,)
+                                && matches!(b.kind, ty::GenericParamDefKind::Lifetime))
+                            || (matches!(a.kind, ty::GenericParamDefKind::Const { .. })
+                                && matches!(b.kind, ty::GenericParamDefKind::Const { .. })))
+                },
+            )
+        };
+
+        if let Some(other_trait_def_id) = self.tcx.all_traits_including_private().find(|def_id| {
+            trait_def_id != *def_id
+                && trait_name == self.tcx.item_name(def_id)
+                && trait_has_same_params(*def_id)
+                && self
+                    .tcx
+                    .predicates_of(*def_id)
```

Comment:
can you explain why you're checking `predicates_of` of the other trait here?

Reply:
I have to pass a Predicate, which owns a TraitRef with the new "trait_defid", to predicate_must_hold_modulo_regions(). So either I remap the existing TraitPredicate with a new TraitRef (which owns the new trait_def_id) or I have to retrieve an existing trait_predicate for the selected trait.

Reply:
`TraitRef` is copy and you can create a new one: `TraitRef::new(new_trait_id, old_trait_predicate.trait_ref.args)` or sth like that

Reply:
Ahhhh... ok I will give it a try. Thanks :)